Don't heap allocate for active requests

hamishknight opened the pull request:
Introduce `ActiveRequest` which can refer to a request on the stack, and can be converted into an `AnyRequest` when we want persistent storage.

hamishknight committed:
Fix Optional's simple_display

hamishknight committed:
[AST] Introduce ActiveRequest

This wrapper is used to type erase a reference to
a request on the stack, and can be converted into
an AnyRequest when persistent storage is required.

Switch the evaluator over to using ActiveRequest
for its stack of active requests, meaning it now
only needs to heap allocate requests when they
enter the cache, or if it's building the (disabled
by default) dependency graph.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test compiler performance

CodaFi commented:
That's a pretty elegant way to avoid allocations on the hot path.  How does this perform without the request graph?

hamishknight commented:
@CodaFi Locally this gives me ~2% improvement in type checking the stdlib with 14% reduction in `Frontend.MaxMallocUsage`, though that gets diluted down significantly when looking at total compile time + memory usage, as we spend most of it in SIL/SILOpt. Though hopefully we won't see such significant dilutions for normal projects.

hamishknight commented:
Jenkins crashed

@swift-ci please test compiler performance

hamishknight committed:
[Evaluator] Stop canonicalizing requests

This is unnecessary given we now only create
AnyRequests when caching, or when building the
dependency graph.

hamishknight commented:
@swift-ci please test compiler performance

hamishknight commented:
@swift-ci please test compiler performance

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
Hmm, not sure what's going on with the request counter fluctuations, as far as I can tell it's unrelated to my change (e.g also observed in https://github.com/apple/swift/pull/29067). The main thing at least is even when they regress, we still see an improvement in the wall time and `Frontend.MaxMallocUsage`.

CodaFi commented on lib/AST/Evaluator.cpp:
```diff
@@ -75,35 +73,40 @@ void Evaluator::emitRequestEvaluatorGraphViz(llvm::StringRef graphVizPath) {
     // If there is an active request, record it's dependency on this request.
     if (!activeRequests.empty())
-      dependencies[activeRequests.back()].push_back(request);
+      dependencies[AnyRequest(activeRequests.back())].push_back(req);
```
It's no big deal for now, but a little micro-optimization here and elsewhere could be using `find_as` to do lookups with ActiveRequests instead of AnyRequests.

hamishknight replied:
@CodaFi Good idea! How does this look?

CodaFi replied:
Fantastic.  Let’s get another CI run and Doug’s rubber stamp.

CodaFi reviewed (approved):
Bravo

hamishknight committed:
Enable the use of find_as with ActiveRequest

This avoids having to allocate a full AnyRequest
for a hash lookup.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

slavapestov commented:
Those request counts fluctuate with every test. We need to investigate why at some point, but I wouldn't worry about it in the context of this PR.

DougGregor reviewed (approved):
Very cool!

compnerd commented:
I believe this may have impacted the Windows build: https://dev.azure.com/compnerd/swift-build/_build/results?buildId=24852&view=logs&j=0e514031-a9b0-5d19-4075-e5711178f10f&t=6491fbec-4801-5529-2943-0955e3c2f90f&l=3804

compnerd commented:
Additional CI: https://ci-external.swift.org/job/oss-swift-windows-x86_64/2565/
